### PR TITLE
Collect extractors into `request` module

### PR DESCRIPTION
**Pull request by hlbarber**
## Motivation and Context

In `axum` there is a definitive module which exports extractors. 

https://docs.rs/axum/latest/axum/extract/index.html

With `pub` `FromParts` being released shortly we might want to collect extractors as well.

## Description

- Move `ConnectInfo` and its associated `Layer` into `request::connect_info`.
- Move `Extension` into `request::extension`.

## Notes

- This is technically not a breaking change as long as it's merged prior to `0.52`. We re-export `Extension` from `extensions` and `ConnectInfo` API has not been released yet.

### CODE REVIEW

File: rust-runtime/aws-smithy-http-server/src/routing/mod.rs [hunk 1]
```diff
@@ -120,18 +116,6 @@ where
...
 
-    /// Convert this router into a [`MakeService`], that is a [`Service`] whose
-    /// response is another service, and provides a [`ConnectInfo`] object to service handlers.
-    ///
-    /// This is useful when running your application with hyper's
-    /// [`Server`].
-    ///
-    /// [`Server`]: hyper::server::Server
-    /// [`MakeService`]: tower::make::MakeService
-    pub fn into_make_service_with_connect_info<C>(self) -> IntoMakeServiceWithConnectInfo<Self, C> {
```
Comment (hlbarber) [inline rust-runtime/aws-smithy-http-server/src/routing/mod.rs:131]: `ConnectInfo` uses the `FromParts` trait which is not functional in the older service builder.

File: rust-runtime/aws-smithy-http-server/src/extension.rs [hunk 2]
```diff
@@ -50,14 +21,12 @@
...
 use thiserror::Error;
 
-use crate::{
-    body::{empty, BoxBody},
-    request::{FromParts, RequestParts},
-    response::IntoResponse,
-};
+use crate::request::RequestParts;
+
+pub use crate::request::extension::Extension;
```
Comment (hlbarber) [inline rust-runtime/aws-smithy-http-server/src/extension.rs:28]: Removing these two re-exports would be a breaking change.

**Comment by hlbarber** [comment]
This should be merged before https://github.com/awslabs/smithy-rs/pull/1886

File: rust-runtime/aws-smithy-http-server/src/request/connect_info.rs [hunk 3]
```diff
@@ -48,12 +50,11 @@ use tower_http::add_extension::{AddExtension, AddExtensionLayer};
 
 use crate::{request::FromParts, Extension};
 
-/// A [`MakeService`] created from a router.
+/// A [`MakeService`] used to insert [`ConnectInfo<T>`] into [`http::Request`]s.
 ///
-/// See [`Router::into_make_service_with_connect_info`] for more details.
```
Comment (hlbarber) [inline rust-runtime/aws-smithy-http-server/src/request/connect_info.rs:53]: This intradoc link would be possible with https://github.com/awslabs/smithy-rs/issues/2014

**Review by 82marbag — APPROVED** [review]

**Pull request merged**